source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,272 @@ namespace Slang
+        {
+            ConstructorDecl* defCtor = nullptr;
+            List<ConstructorDecl*> ctorList;
+        };
+
+        // for some reason `Dictionary` corrupts stack when using large types as a value in my use-case.
+        // Due to this, `AggTypeDeclWithCtorData*` is data from `inheritanceData`
+        Dictionary<AggTypeDecl*, AggTypeDeclWithCtorData*> inheritanceDictToData;
+        List<AggTypeDecl*> linearInheritanceGraph;
+        List<AggTypeDeclWithCtorData> inheritanceData;
+
+        void addDecl(AggTypeDecl* decl)
+        {
+            linearInheritanceGraph.add(decl);
+            if (inheritanceDictToData.containsKey(decl))
+                return;
+
+            inheritanceData.add({});
+            inheritanceDictToData[decl] = &inheritanceData.getLast();
+        }
+    };
+
+    static Facet::DirectnessVal _getAuxDataFromInheritanceInfo(LinearInheritanceGraphWithAuxInfo& linearInheritanceGraph, InheritanceInfo& inheritanceInfo, Facet::DirectnessVal maxDepthToStore)
+    {
+        // Returns LinearInheritanceGraphWithInfo ordered "this"->"super".
+        // We will assume multiple `InheritanceDecl`s are possible. We also will assume equal
+        // `InheritanceDecl`s may show up multiple times inside the inheritance-graph.
+        // This is required to get proper inheritance depth and for setting up the inheritance
+        // graph for data processing on non-duplicated data
+
+        Facet::DirectnessVal inheritanceDepth = 0;
+        for (const auto& facet : inheritanceInfo.facets)
+        {
+            inheritanceDepth = (inheritanceDepth > facet->directness) ? inheritanceDepth : facet->directness;
+
+            auto facetDecl = facet->getDeclRef().getDecl();
+            auto aggTypeDecl = as<StructDecl>(facetDecl);
+            if (!aggTypeDecl)
+                continue;
+
+            LinearInheritanceGraphWithAuxInfo::AggTypeDeclWithCtorData* aggTypeInfo{};
+            if (inheritanceDepth <= maxDepthToStore)
+            {
+                linearInheritanceGraph.addDecl(aggTypeDecl);
+                aggTypeInfo = linearInheritanceGraph.inheritanceDictToData[aggTypeDecl];
+            }
+            else
+                continue;
+
+            for (auto m : aggTypeDecl->members)
+            {
+                // Forward declared constructors may be declared in an interface, these must be ignored.
+                // We can determine these based on the body being nullptr
+                auto ctor = as<ConstructorDecl>(m);
+                if (!ctor || !ctor->body)
+                    continue;
+                aggTypeInfo->ctorList.add(ctor);
+                if (ctor->members.getCount() != 0)
+                    continue;
+                aggTypeInfo->defCtor = ctor;
+            }
+        }
+        return inheritanceDepth;
+    }
+
+    static ConstructorDecl* _createCtor(SemanticsDeclAttributesVisitor* visitor, ASTBuilder* m_astBuilder, AggTypeDecl* decl)
```

Comment:
Why is this done in SemanticsAttributeVisitor? I feel the ctor signature synthesis should be done earlier in SemanticsDeclHeaderVisitor, and the ctor body synthesis can be done later in SemanticsDeclBodyVisitor.

Reply:
The code was done in `SemanticsAttributeVisitor` to ensure we created CTor the earliest possible after type resolution. 

Splitting should work, all we need is the ability to delete a ctor if unused + access to parent member list.

I suppose we can:
1. Create a default ctor if missing in `SemanticsDeclHeaderVisitor`.
2. During `SemanticsDeclBodyVisitor` we can fill the body, if it won't be filled we delete.